add KVS checkpoints

garlick opened the pull request:
Thought I'd park this although it needs further work.

This adds a table to the content-sqlite database that can store arbitrary key-value pairs, then modifies the KVS to read its root blobref from this table (if available) during startup, and write its final root blobref during shutdown.  In the future, the table could accommodate multiple keys for live namespaces if we want to go that way (but right now we don't need to support that).

If an instance has the `content.backing-path` attribute set to the path to a sqlite file when the content-sqlite module is loaded, then it is used (existing content preserved).  If it is set but the file does not exist, it is created empty, and retained on exit.

The net result is the systemd unit file _could_ set this attribute to say `/var/lib/flux/content.sqlite` and then the KVS is preserved across service restarts.

I got rid of the `persist-filesystem` and `persist-directory` attributes.  Hopefully they will not be missed.

Couple loose ends
* If the hash method changes (from sha1 to sha256 say) this is not handled
* Need to check but I think sha256 hash digests are being truncated to sha1 size right now in the backing store, which would be a big gaping security hole for multi-user
* Maybe it would be nice to have control over "restart" behavior independent of the path setting
* Ultimately this should be handled via config not attributes, but this was easier to get launched
* More testing required

garlick commented:
Rebased on current master and forced a push.

grondo commented:
Without actually looking at the changes :stuck_out_tongue:, does this make it impossible for multiple instances to share the same content backing store? Not sure if that was even a use case, but it felt like before this was actually a possibility, but with non-unique metadata, no longer so.

(Sorry if a dumb question)

garlick commented:
Thanks @chu11 for the review comments, really helpful!  I'll fix those tomorrow.

> does this make it impossible for multiple instances to share the same content backing store? Not sure if that was even a use case, but it felt like before this was actually a possibility, but with non-unique metadata, no longer so.

Uhh yeah.  True.   I don't think it's a near term use case, but we do have a nice clean design that permits it, and this messes it up.  Let me ponder whether there's a way we can do this that's looks less like a malignant growth on a perfectly good abstraction.

garlick commented on src/broker/broker.c:
```diff
@@ -149,7 +149,6 @@ static void runlevel_cb (runlevel_t *r, int level, int rc, double elapsed,
 static void runlevel_io_cb (runlevel_t *r, const char *name,
                             const char *msg, void *arg);
 
-static int create_persistdir (attr_t *attrs, uint32_t rank);
```
I realized I spread the removal of this over several commits.  I'll try to consolidate and include a good, searchable commit message.

garlick commented:
Just forced a push that refactors/rewords some of the commits and hopefully addresses all @chu11's review comments.

I went ahead and implemented a separate service called `kvs-checkpoint` for the checkpoint metadata.  It is still implemented in the `content-sqlite` module at the moment, to share the same database, but I've tried to indicate that it's not part of the content store per se, and that this was done for expediency.  Also, I removed the metaget/metaput subcommands from `flux-content(1)` and instead just implemented those in a test program since really this wasn't intended for general purpose use.  Hopefully that at least makes its mixing with the content-sqlite service less of a design violation.

garlick commented on src/cmd/builtin/content.c:
```diff
@@ -207,6 +277,20 @@ static struct optparse_subcommand content_subcmds[] = {
+      NULL,
+    },
+    { "metaput",
+      "KEY VAL",
```
I ended up dropping this from `flux-content(1)` and moving to a test program.  I retained the existing usage but I assume it's no big deal in a test program.

garlick commented:
This is probably ready for a review.  I might be able to increase the diff coverage in a couple of cases, but a lot of the uncovered code is in hard to reach error paths.

garlick commented on src/modules/content-sqlite/content-sqlite.c:
```diff
@@ -376,27 +375,31 @@ static void content_sqlite_closedb (struct content_sqlite *ctx)
             if (sqlite3_finalize (ctx->store_stmt) != SQLITE_OK)
                 log_sqlite_error (ctx, "sqlite_finalize store_stmt");
         }
-        if (ctx->load_stmt) {
```
Nice catch!  Fixing...

chu11 commented:
Good to go now?  I'll let ya set merge-when-passing if ya think so.

garlick commented:
Thanks!

Restarted a builder that failed like this:
```
There was an error in the .travis.yml file from which we could not recover.
Unfortunately, we do not know much about this error.
Please review https://docs.travis-ci.com, or contact us at support@travis-ci.com with the error ID: cee1e4cd24984025927555fb39ff1d89
```
